drivers/rtc/rtc_nxp_irtc.c

```diff
@@ -0,0 +1,454 @@
+	    (time_value->tm_mon < 0 || time_value->tm_mon > 11)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_YEAR)) &&
+	    ((time_value->tm_year - RTC_NXP_IRTC_YEAR_OFFSET) < -128 ||
+	     (time_value->tm_year - RTC_NXP_IRTC_YEAR_OFFSET) > 127)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_WEEKDAY)) &&
+	    (time_value->tm_wday < -1 || time_value->tm_wday > 6)) {
+		return 0;
+	}
+
+	if ((mask == 0) && (time_value->tm_isdst < -1 || time_value->tm_isdst > 1)) {
+		return 0;
+	}
+
+	return 1;
+}
+
+/*
+ * The runtime where this is accessed is unknown so we force a lock on the registers then force an
+ * unlock to guarantee 2 seconds of write time
+ */
+static void nxp_irtc_unlock_registers(RTC_Type *reg)
+{
+	/* Lock the regsiters */
+	while ((reg->STATUS & (uint16_t)RTC_STATUS_WRITE_PROT_EN_MASK) == 0) {
+		*(uint8_t *)(&reg->STATUS) |= RTC_STATUS_WE(0x2);
+	}
+	/* Unlock the registers */
+	while ((reg->STATUS & (int16_t)RTC_STATUS_WRITE_PROT_EN_MASK) != 0) {
+		/*
+		 * The casting is required for writing only a single Byte to the STATUS Register,
+		 * the pattern here unlocks all RTC registers for writing. When unlocked if a 0x20
+		 * if written to the STATUS register the RTC registers will lock again and the next
+		 * write will lead to a fault.
+		 */
+		*(volatile uint8_t *)(&reg->STATUS) = 0x00;
+		*(volatile uint8_t *)(&reg->STATUS) = 0x40;
+		*(volatile uint8_t *)(&reg->STATUS) = 0xC0;
+		*(volatile uint8_t *)(&reg->STATUS) = 0x80;
+	}
+}
+
+static int nxp_irtc_set_time(const struct device *dev, const struct rtc_time *timeptr)
+{
+	const struct nxp_irtc_config *config = dev->config;
+	struct nxp_irtc_data *data = dev->data;
+	RTC_Type *irtc_reg = config->base;
+
+	if (!timeptr || !is_rtc_time_in_bounds(timeptr, 0)) {
+		return -EINVAL;
+	}
+
+	int calc_year = timeptr->tm_year - RTC_NXP_IRTC_YEAR_OFFSET;
+	/* The IRTC Month Index starts at 1 instead of 0 */
+	int calc_month = timeptr->tm_mon + 1;
+
+	uint32_t key = irq_lock();
+
+	nxp_irtc_unlock_registers(irtc_reg);
+	irtc_reg->SECONDS = RTC_SECONDS_SEC_CNT(timeptr->tm_sec);
+
+	irtc_reg->HOURMIN = RTC_HOURMIN_MIN_CNT(timeptr->tm_min) |
+		RTC_HOURMIN_HOUR_CNT(timeptr->tm_hour);
+
+	/* 1 is valid for rtc_time.tm_wday property but is out of bounds for IRTC registers */
+	irtc_reg->DAYS = RTC_DAYS_DAY_CNT(timeptr->tm_mday) |
+		      (timeptr->tm_wday == -1 ? 0 : RTC_DAYS_DOW(timeptr->tm_wday));
+
+	irtc_reg->YEARMON = RTC_YEARMON_MON_CNT(calc_month) | RTC_YEARMON_YROFST(calc_year);
+
+	if (timeptr->tm_isdst != -1) {
+		irtc_reg->CTRL |= RTC_CTRL_DST_EN(timeptr->tm_isdst);
+		data->is_dst_enabled = true;
+	}
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+static int nxp_irtc_get_time(const struct device *dev, struct rtc_time *timeptr)
+{
+	const struct nxp_irtc_config *config = dev->config;
+	struct nxp_irtc_data *data = dev->data;
+	RTC_Type *irtc_reg = config->base;
+
+	if (!timeptr) {
+		return -ENODATA;
```

Comment:
Just like the `*dev` arg, the `*timeptr` is required, so you don't need to check it :) save some ROM and remove the check :)

Reply:
I can still remove the check but I only added this since the rtc.h file stated this as a possible return value, and this was all I could figure the rtc.h description was referring to. Let me know what you think.

Reply:
yeah, the point is that if the `timeptr` is NULL, a NULL dereference occurs, which causes a hard fault, so it will never return :) No return value required for that case :D

Reply:
Code replaced with an assert